### PR TITLE
Remove compat javascript behaviour for string concatenation

### PR DESCRIPTION
**Pull request by MaxGraey**
This pull request removes JS  behaviour for null-sh strings like this:

**Before (master)**
```ts
const path = "/dev/" + null; // actually needs changetype<string>(null)
// > "/dev/null"
```
**After (this PR)**
```ts
const path = "/dev/" + null;
// > "/dev/"
```

Also removed null checks for concat, startsWith and endsWith.

- [x] I've read the contributing guidelines

### CODE REVIEW

File: std/assembly/string.ts [hunk 1]
```diff
@@ -80,11 +80,11 @@ import { Array } from "./array";
   }
 
   @operator("+") private static __concat(left: String, right: String): String {
-    return select<String>(left, changetype<String>("null"), left !== null).concat(right);
+    return left !== null ? left.concat(right) : right;
   }
 
   concat(other: String): String {
-    if (other === null) other = changetype<String>("null");
+    if (other === null) return this;
```
Comment (dcodeIO) [inline std/assembly/string.ts:87]: Iirc this case is left over from the early days, just in case. Perhaps we can remove the line entirely nowadays?

File: std/assembly/string.ts [hunk 2]
```diff
@@ -196,7 +196,7 @@ import { Array } from "./array";
   }
 
   startsWith(search: String, start: i32 = 0): bool {
-    if (search === null) search = changetype<String>("null");
+    if (search === null) return false;
```
Comment (dcodeIO) [inline std/assembly/string.ts:199]: Similar to the line in `.concat` I think.

File: std/assembly/string.ts [hunk 3]
```diff
@@ -80,11 +80,11 @@ import { Array } from "./array";
   }
 
   @operator("+") private static __concat(left: String, right: String): String {
-    return select<String>(left, changetype<String>("null"), left !== null).concat(right);
+    return left !== null ? left.concat(right) : right;
```
Comment (dcodeIO) [inline std/assembly/string.ts:83]: This one could be a little tricky, depending on how the overload behaves. We may be able to remove the check here as well, but then should double check that the compiler never allows `null` values when compiling the overload. Something along the lines of

```ts
var a: string = "a";
var b: string | null = "b";
a + b // should complain
b + a // should complain
````

File: std/assembly/string.ts (hunk 1 above)
Comment (MaxGraey) [reply std/assembly/string.ts:87]: I'm not sure about this. In theory still may got situation when `other` is `null` with non-nullish type. For example when this string comes from host.

Comment (MaxGraey) [reply std/assembly/string.ts:87]: another variant:
```ts
var a: string = "a";
var b: string | null = "b";
a + b
```

File: std/assembly/string.ts [hunk 4]
```diff
@@ -79,12 +79,13 @@ import { Array } from "./array";
     return (first - 0xD800 << 10) + (second - 0xDC00) + 0x10000;
   }
 
-  @operator("+") private static __concat(left: String, right: String): String {
-    return select<String>(left, changetype<String>("null"), left !== null).concat(right);
+  @operator("+") private static __concat(left: String | null, right: String | null): String {
+    if (right === null) return select<String>(left!, changetype<String>(""), left !== null);
+    if (left === null) return right;
+    return left.concat(right);
```
Comment (MaxGraey) [inline std/assembly/string.ts:85]: So this guarantee `a + b` will always produce non-nullabe string.

```ts
let a: string | null = 'a';
let b: string | null = 'b';
a + b // 'ab'

b = null;
a + b // 'a'

a = null;
b = 'b';
a + b // 'b'

a = null;
b = null;
a + b; // '' 
```

Comment (MaxGraey) [reply std/assembly/string.ts:85]: Another option is making this more stricter and accept only non-nullable strings for unary/binary string operations. In this case we can't write following anymore:
```ts
let a: string | null = 'a';
let b: string | null = 'b';
let c = a + b; // AS Error
```
and will use something like this:
```ts
let a: string | null = 'a';
let b: string | null = 'b';
let c = a! + b!;
```

File: std/assembly/string.ts (hunk 1 above)
Comment (dcodeIO) [reply std/assembly/string.ts]: Yeah, it is possible to construct such a situation with an export, yet I wonder whether guarding for invalid data from external sources at all the places is desirable as we also want stuff to be efficient internally. If I'm not mistaken we are not guarding for this anywhere else, just here, so unless we want to establish that convention everywhere in stdlib, the alternative is to not bother, even not here.

Comment (MaxGraey) [reply std/assembly/string.ts]: I removed this checks

File: std/assembly/string.ts (hunk 4 above)
Comment (dcodeIO) [reply std/assembly/string.ts:85]: Yeah, that's the two options we have. My intuition currently is that if we allow `a + b` with one potentially `null`, we should deal with that like JS does, i.e. concatenating with the string `"null"`. It's fine to only do it in the operator overload, though, where it can happen, but we can remove the check everywhere else where the type is non-nullable anywhere, as mentioned in the previous comments.

Comment (MaxGraey) [reply std/assembly/string.ts:85]: > My intuition currently is that if we allow a + b with one potentially null, we should deal with that like JS does, i.e. concatenating with the string "null"

I agree we should deal with nullish strings. But not sure about JS-way which convert null to "null" string. This is contrary to the general philosophy of AssemblyScript so it seems to me =) So I suggest have one special case for `<string>null + <string>null == ""` which looks more unexpected than "nullnull". See https://github.com/AssemblyScript/assemblyscript/pull/1682#discussion_r575676957

Comment (dcodeIO) [reply std/assembly/string.ts:85]: I'm torn on this tbh. On the one hand what you suggest makes sense because concatenating with `null` doesn't really make a lot of sense anyhow, but on the other that'd be another deviation from JS behavior, and seems premature to do just because why not. I somewhat tend to not risk this to become a problem, or another peculiarity in the docs, and keep JS semantics in tact fwiw, that is unless we have a really compelling reason to change it.

Comment (dcodeIO) [reply std/assembly/string.ts:85]: Perhaps an intuitive way to think of these things is in terms of "your fix may be someone else's bug" :)

Comment (MaxGraey) [reply std/assembly/string.ts:85]: I thought this good reason (especially due to we support wasi):
```ts
const filename: string | null = null;
const path = "/dev/" + filename;
// > "/dev/null"
```

Comment (MaxGraey) [reply std/assembly/string.ts:85]: > Perhaps an intuitive way to think of these things is in terms of "your fix may be someone else's bug" :)

C# support this:
```c#
string bob = "abc" + null + null + null + "123";  // abc123
```

Comment (dcodeIO) [reply std/assembly/string.ts:85]: Hmmhmhm. I guess we may benefit from collecting a few opinions about this, and see what others would prefer in situations like these. Would you be interested in bringing this up in the next meeting, elaborate a bit on the what, how and why? :)

Comment (MaxGraey) [reply std/assembly/string.ts:85]: Yeah, sure! Let's discuss it at the next meeting

**Comment by MaxGraey** [comment]
So anther approach suggested by @vgrichina is throw exception for null strings during concat operation. But in this case we probably should propagate this strict behaviour for rest strings methods as well. Thought?

**Comment by saulecabrera** [comment]
I had to think about this a bit before voicing my opinion. I believe that in the end, all of this boils down to consistency as @battlelinegames mentioned yesterday in the meeting; if one of the principles of the language is to be as close as possible to JS, then as a user I'd expect these quirks to be there, especially when other typed languages that target JS like TS keep this behavior too. Continuing with the consistency line of thought, if we decide on removing it, then we should clearly state under what principle we decided on removing it, and probably the sanest thing to do moving forward is to apply that "improvement" principle to all the other JS quirks (where possible obviously). In reality, though, _improving JS_ could represent breaking changes, but such discussion falls again under the bucket of what level of compatibility we want to have with it.

TLDR; there's no doubt that this behavior can be improved, and people will vote or share their opinions based on _how_ that improvement can be made, but I think that the bigger discussion here is about the principles of the language rather than how to approach the actual feature.

**Comment by MaxGraey** [comment]
@dcodeIO could you review this?

**Comment by dcodeIO** [comment]
Is it correct to assume, from the diff, that if there is any breaking change in the PR it is in the `__gt` signature, which is more like a bugfix since `__lt` doesn't have nullable types?

**Comment by MaxGraey** [comment]
Yes I guess it's breaking change due to some of signatures was changed

**Comment by dcodeIO** [comment]
Considering to label this one a "fix", as it merely fixes that one signature that is different from all the others. Wdyt?

**Comment by MaxGraey** [comment]
Up to you. I don't think it should significantly affect to existing user's code. So it could be marked as `fix` or `feat` (improvments/simplification)
